[DOCS] Adds regression example (#650)
elastic/stack-docs: docs/en/stack/ml/df-analytics/flightdata-regression.asciidoc

[role="xpack"]
[testenv="platinum"]
[[flightdata-regression]]
=== Predicting flight delays with {reganalysis}

Let's try to predict flight delays by using the 
{kibana-ref}/add-sample-data.html[sample flight data]. We want to be able to use 
information such as weather and location of the destination and origin, flight 
distance and carrier to predict the number of minutes delayed for each flight. 
As it is a continuous numeric variable, we'll use {reganalysis} to make the 
prediction.

This dataset consists of sample data that has been manually created and it 
contains inconsistencies. For example, a flight can be both delayed and 
canceled. Nonetheless, the use case is relevant and it is a good example of how 

> the quality of input data will affect the quality of results.

Reviewer: The tone of the para does not sit quite right, possibly a bit arrogant/unfriendly. Can we soften this please. Perhaps ...

```
We have chosen this dataset as an example because it is easily accessible
for Kibana users and the use case is relevant. However the data has been manually 
created and contains some inconsistencies. For example, a flight can be both delayed 
and canceled. Please remember that the quality of your input data will affect the quality
of results.
```

    "OriginAirportID": "ADL",
    "OriginRegion": "SE-BD",
    "DestCityName": "Tokoname",
    "FlightTimeHour": 7.577903786991782,
    "FlightDelayMin": 0
  }
}
```

{regression-cap} is a supervised machine learning analysis and therefore needs 
to train on data that contains the ground truth for the `dependent_variable` 
that we want to predict. In this example, the ground truth is available in each 
document as the actual value of `FlightDelayMins`. In order to be analyzed, a 
document must contain at least one field with a supported data type (`numeric`, 
`boolean`, `text`, `keyword` or `ip`) and must not contain arrays with more than 
one item.

If your source data consists of some documents that contain a 
`dependent_variable` and some that do not, the model is trained on the 
`training_percent` of the documents that contain ground truth. However, 
predictions are made against all of the data. The current implementation of 
{reganalysis} supports a single batch analysis for both training and 
predictions.

[[flightdata-regression-model]]
==== Creating a {regression} model

To predict the number of minutes delayed for each flight:
. Create a {dfanalytics-job}.

> +

Reviewer: This `+` character displays in preview... not sure if this is a typo or not. It does not display as the first item in a numbered list `1.`  .. this might just be a problem with preview...

    ],
    "query": { <2>
      "range": {
        "DistanceKilometers": { 
          "gt": 0
        }
      }
    }
  },
  "dest": {
    "index": "df-flight-delays"  <3>
  },
  "analysis": {
    "regression": {
      "dependent_variable": "FlightDelayMin",  <4>
      "training_percent": 90  <5>
    }
  },
  "analyzed_fields": {
    "includes": [],
    "excludes": [    <6>
      "Cancelled",
      "FlightDelay",
      "FlightDelayType"
    ]
  },
  "model_memory_limit": "100mb" <7>
}
--------------------------------------------------
// TEST[skip:setup kibana sample data]

<1> The source index to analyze.
<2> This query removes erroneous data from the analysis to improve its quality.
<3> The index that will contain the results of the analysis; it will consist of 
a copy of the source index data where each document is annotated with the 
results.
<4> Specifies the continuous variable we want to predict with the {reganalysis}.
<5> Specifies the proportion of data that is used for training the model. In 
this example we use 90% of the source data. The training data is randomly 
selected and provide predictions from the remaining 10%. These percentages are 

> approximate.

Reviewer: Predictions are made for all of he data, not just the remaining 10%. Therefore should be:

```
Specifies the approximate proportion of data that is used for training the model. In 
this example we randomly select 90% of the source data for training. 
```

<6> Specifies fields to be excluded from the analysis. It is recommended to 
exclude fields that either contain erroneous data or describe the 
`dependent_variable`.
<7> Specifies a memory limit for the job. If the job requires more than this 
amount of memory, it fails to start. This makes it possible to prevent job 
execution if the available memory on the node is limited.
--

. Start the job.
+
--
Use the {ref}/start-dfanalytics.html[start {dfanalytics-jobs}] API to start the 
job. It will stop automatically when the analysis is complete, you don't need to 

> stop it manually.

Reviewer: nit. Saying the same thing twice in the same sentence.

[role="xpack"]
[testenv="platinum"]
[[flightdata-regression]]
=== Predicting delays with {reganalysis} by using Kibana sample flight data

> The goal of <<dfa-regression,{reganalysis}>> is to find relationships among 

Reviewer: This first para does not really add much and lowers the energy level into the page. Thoughts?

Author: @sophiec20 I tried to follow the pattern of the other example in this chapter (outlier detection) which starts with a short definition of the analysis type. I think it is necessary to emphasize in the beginning what this example is about. It seems like a natural introduction to the topic. I play with the order and the wording and fix it in my next commit.

https://www.elastic.co/guide/en/elastic-stack-overview/current/ecommerce-outliers.html

different features of your data, then make predictions on the new data points 
based on the previously found relationships. 

Let's try to predict flight delays by using the 
{kibana-ref}/add-sample-data.html[sample flight data]. We use this data to learn 
a model that allows us to predict flight delays based on information such as 
weather and location of the destination and origin, flight distance and carrier. 
We'd like to predict the number of minutes delayed for each flight. As it is a 
continuous numeric variable, we'll use {reganalysis} to make the prediction.

This dataset consists of sample data that has been manually created and it 
contains inconsistencies. For example, a flight can be both delayed and 
canceled. However, it is easy to set up in Kibana and the use case is 
applicable, so it is suitable to be an example. Still, don't forget this is only 

> sample data, do not expect adequate results.

Reviewer: For added insight, suggest

`Despite the data inconsistencies, flight delays can be used to demonstrate regression analysis because
it is easy to set up in Kibana, the use case is relevant and it is a good example of how the 
quality of input data will affect the quality of results.`

> sample data, do not expect adequate results.

Reviewer: For added insight, suggest something along the lines of...

`Despite the data inconsistencies, flight delays can be used to demonstrate regression analysis because
it is easy to set up in Kibana, the use case is relevant and it is a good example of how the 
quality of input data will affect the quality of results.`

We will create an {dfanalytics-job} to predict the minutes that flights are 
delayed by. Each document in the dataset contains details for a single flight, 
so this data is ready for analysis as it is already in a two-dimensional 
entity-based data structure that we call a {dataframe}. Depending on your source 
data, it may be required to first {ref}/transforms.html[transform] your source 
data into an entity-centric index before you could setup and start the analysis.

This is an example source document from the dataset:

```

> {"_index":"kibana_sample_data_flights","_type":"_doc","_id":"S-JS1W0BJ7wufFIaPAHe","_version":1,"_seq_no":3356,"_primary_term":1,"found":true,"_source":{"FlightNum":"N32FE9T","DestCountry":"JP","OriginWeather":"Thunder & Lightning","OriginCityName":"Adelaide","AvgTicketPrice":499.08518599798685,"DistanceMiles":4802.864932998549,"FlightDelay":false,"DestWeather":"Sunny","Dest":"Chubu Centrair International Airport","FlightDelayType":"No Delay","OriginCountry":"AU","dayOfWeek":3,"DistanceKilometers":7729.461862731618,"timestamp":"2019-10-17T11:12:29","DestLocation":{"lat":"34.85839844","lon":"136.8049927"},"DestAirportID":"NGO","Carrier":"ES-Air","Cancelled":false,"FlightTimeMin":454.6742272195069,"Origin":"Adelaide International Airport","OriginLocation":{"lat":"-34.945","lon":"138.531006"},"DestRegion":"SE-BD","OriginAirportID":"ADL","OriginRegion":"SE-BD","DestCityName":"Tokoname","FlightTimeHour":7.577903786991782,"FlightDelayMin":0}}

Reviewer: It was difficult to see this document in the preview due to long horiz scroll. Is it possible to view as line wrapped JSON?

to train on data that contains the ground truth for the `dependent_variable` 
that we want to predict. In this example, the ground truth is available in each 
document as the actual value of `FlightDelayMins`.

If your source data consisted of documents that contains a `dependent_variable` 
and some that does not contain a `dependent_variable`, then training is 
performed on the `training_percent` of the documents that contain ground truth 
and predictions will be made against all of the data. The current implementation 
of {reganalysis} supports a single batch analysis for both training and 
predictions. Future versions will allow a learned model to be incorporated at 
index time.

. Create a {dfanalytics-job}.
+
--
Use the {ref}/put-dfanalytics.html[create {dfanalytics-jobs}] API as you can see 
in the following example:

[source,console]
--------------------------------------------------
PUT _ml/data_frame/analytics/model-flight-delays
{
  "source": {
    "index": [
      "kibana_sample_data_flights" <1>
    ],
    "query": { <2>
      "range": {
        "DistanceKilometers": { 
          "gt": 0
        }
      }
    }
  },
  "dest": {
    "index": "df-flight-delays",  <3>

>     "results_field": "ml" 

Reviewer: `results_field` is a default value. It can be removed from the Create statement (as it will show up in the response anyway. Slight simplification.

  },
  "analysis": {
    "regression": {
      "dependent_variable": "FlightDelayMin",  <4>
      "training_percent": 90  <5>
    }
  },
  "analyzed_fields": {
    "includes": [],
    "excludes": [    <6>
      "Cancelled",
      "FlightDelay",
      "FlightDelayType"
    ]
  },
  "model_memory_limit": "100mb" <7>
}
--------------------------------------------------
// TEST[skip:setup kibana sample data]

<1> The source index to analyze.
<2> This query removes erroneous data from the analysis to improve its quality.

> <3> The index that contains the results of the analysis; it consists of a copy 

Reviewer: Should this be in future tense. i.e. the index that will contain the results of analysis ??

Author: As far as I know, we use present simple tense in these explanations. Still, I change it to future tense to indicate that the destination index will be the output of the analysis.

exclude fields that either contain erroneous data or describe the 
`dependent_variable`.
<7> Specifies a memory limit for the job. If the job requires more than this 
amount of memory, it fails to start. This makes it possible to prevent job 
execution if the available memory on the node is limited.

The API returns the following response:

[source,console-result]
--------------------------------------------------  
{
  "id" : "model-flight-delays",
  "source" : {
    "index" : [
      "kibana_sample_data_flights"
    ],
    "query" : {
      "range" : {
        "DistanceKilometers" : {
          "gt" : 0
        }
      }
    }
  },
  "dest" : {
    "index" : "df-flight-delays",
    "results_field" : "ml"
  },
  "analysis" : {
    "regression" : {
      "dependent_variable" : "FlightDelayMin",
      "training_percent" : 90.0
    }
  },
  "analyzed_fields" : {
    "includes" : [ ],
    "excludes" : [
      "Cancelled",
      "FlightDelay",
      "FlightDelayType"
    ]
  },
  "model_memory_limit" : "100mb",
  "create_time" : 1571918329132,
  "version" : "7.5.0",
  "allow_lazy_start" : false
}
--------------------------------------------------
--

. Start the job.
+
--

> Use the {ref}/start-dfanalytics.html[start {dfanalytics-jobs}] and 

Reviewer: If we mention `_stop` here, we should also mention earlier that they will not need to stop this analysis as it will stop itself once the analysis is complete.

documents, the longer the job to run.
--

. Check the job stats to follow the progress by using the 
{ref}/get-dfanalytics-stats.html[get {dfanalytics-jobs} statistics API].
+
--

[source,console]
--------------------------------------------------
GET _ml/data_frame/analytics/model-flight-delays/_stats
--------------------------------------------------
// TEST[skip:TBD]

The API call returns the following response: 

[source,console-result]
----  
{
  "count" : 1,
  "data_frame_analytics" : [
    {
      "id" : "model-flight-delays",
      "state" : "stopped",
      "progress" : [
        {
          "phase" : "reindexing",
          "progress_percent" : 100
        },
        {
          "phase" : "loading_data",
          "progress_percent" : 100
        },
        {
          "phase" : "analyzing",
          "progress_percent" : 100
        },
        {
          "phase" : "writing_results",
          "progress_percent" : 100
        }
      ]
    }
  ]
}
----  

The job has four phases. When all the phases have completed, the job 
state becomes `stopped` and the results are ready to view and evaluate.
--

[[flightdata-regression-results]]
==== Viewing results

> . Run the following command to view the results in the destination index:

Reviewer: I think it is worth mentioning that the results are stored in a "normal" elasticsearch index.
e.g. `Use the standard elasticsearch search command to view the results in the destination index`

The results can be evaluated for documents which contain both the ground truth 
field and the prediction. In the example below, `FlightDelayMins` contains the 
ground truth and the prediction is stored as `ml.FlightDelayMin_prediction`.

. Use the {dfanalytics} evaluate API to evaluate the results.
+
--
First, we want to know the training error that represents how well the model 
performed on the training dataset:

[source,console]
--------------------------------------------------
POST _ml/data_frame/_evaluate
{
 "index": "df-flight-delays",   <1>
  "query": {
      "bool": {
        "filter": [{ "term":  { "ml.is_training": true } }]  <2>
      }
    },
 "evaluation": {
   "regression": {
     "actual_field": "FlightDelayMin",   <3>
     "predicted_field": "ml.FlightDelayMin_prediction", <4>
     "metrics": {  
       "r_squared": {},
       "mean_squared_error": {}                            
     }
   }
 }
}
--------------------------------------------------
// TEST[skip:TBD]

<1> The destination index which is the output of the analysis job.
<2> We calculate the training error by only evaluating the training data.
<3> The ground truth label.
<4> Predicted value.

> You can also evaluate the data that has not been used to train the model, the 

Reviewer: We should give similar emphasis on evaluating training and generalization errors. E.g. something along the lines of...

`Next, we calculate the generalization error that represents how well the model performed on previously unseen data.`

[role="xpack"]
[testenv="platinum"]
[[flightdata-regression]]

> === Predicting delays with {reganalysis} by using Kibana sample flight data

Reviewer 2: Not a big deal, but here's an alternative suggestion for the title
```suggestion
=== Predicting flight delays with {reganalysis}
```

Let's try to predict flight delays by using the 
{kibana-ref}/add-sample-data.html[sample flight data]. We use this data to learn 

> a {regression} model that allows us to predict flight delays based on 

Reviewer 2: Another minor suggestion:
> We use this data to learn a regression model that allows us....

Maybe it's clearer to say something like this:
> We can use {ml} {dfanalytics} to create a regression model that enables us to

information such as weather and location of the destination and origin, flight 

> distance and carrier. We'd like to predict the number of minutes delayed for 

Reviewer 2: ```suggestion
distance and carrier. For example, we can predict the number of minutes delayed for 
```

> each flight. As it is a continuous numeric variable, we'll use {reganalysis} to 

Reviewer 2: > ... we'll use {reganalysis}...

Since we're only deciding to use regression analysis at this point in the procedure, does this mean it was too soon to mention a "regression model" in an earlier sentence?

Author: I think at this point we explain why we use regression analysis instead of anything else, in this regard, it is not the moment of decision.

make the prediction.

This dataset consists of sample data that has been manually created and it 
contains inconsistencies. For example, a flight can be both delayed and 

> canceled. Despite the data inconsistencies, flight delays can be used to 

Reviewer 2: I think this sentence can be shortened:
```suggestion
canceled. 
```

> demonstrate {reganalysis} because it is easy to set up in Kibana, the use case 

Reviewer 2: Suggestion for shortening:
```suggestion
Nonetheless, the use case 
```

is relevant and it is a good example of how the quality of input data will 
affect the quality of results.

> We will create an {dfanalytics-job} to predict the minutes that flights are 

Reviewer 2: ```suggestion
We will create a {dfanalytics-job} to predict the flights delays in minutes. 
```

> delayed by. Each document in the dataset contains details for a single flight, 

Reviewer 2: ```suggestion
Each document in the dataset contains details for a single flight, 
```

> so this data is ready for analysis as it is already in a two-dimensional 

Reviewer 2: ```suggestion
so this data is ready for analysis; it is already in a two-dimensional 
```

> entity-based data structure that we call a {dataframe}. Depending on your source 

Reviewer 2: ```suggestion
entity-based data structure (also known as a _{dataframe}_). Depending on your source 
```

> data, it may be required to first {ref}/transforms.html[transform] your source 

Reviewer 2: ```suggestion
data, you might need to {ref}/transforms.html[transform] the
```

> data into an entity-centric index before you could setup and start the analysis.

Reviewer 2: ```suggestion
data into an entity-centric index before you analyze the data.
```

    "AvgTicketPrice": 499.08518599798685,
    "DistanceMiles": 4802.864932998549,
    "FlightDelay": false,
    "DestWeather": "Sunny",
    "Dest": "Chubu Centrair International Airport",
    "FlightDelayType": "No Delay",
    "OriginCountry": "AU",
    "dayOfWeek": 3,
    "DistanceKilometers": 7729.461862731618,
    "timestamp": "2019-10-17T11:12:29",
    "DestLocation": {
      "lat": "34.85839844",
      "lon": "136.8049927"
    },
    "DestAirportID": "NGO",
    "Carrier": "ES-Air",
    "Cancelled": false,
    "FlightTimeMin": 454.6742272195069,
    "Origin": "Adelaide International Airport",
    "OriginLocation": {
      "lat": "-34.945",
      "lon": "138.531006"
    },
    "DestRegion": "SE-BD",
    "OriginAirportID": "ADL",
    "OriginRegion": "SE-BD",
    "DestCityName": "Tokoname",
    "FlightTimeHour": 7.577903786991782,
    "FlightDelayMin": 0
  }
}
```

{regression-cap} is a supervised machine learning analysis and therefore needs 
to train on data that contains the ground truth for the `dependent_variable` 
that we want to predict. In this example, the ground truth is available in each 
document as the actual value of `FlightDelayMins`.

> If your source data consisted of documents that contains a `dependent_variable` 

Reviewer 2: ```suggestion
If your source data consists of some documents that contain a `dependent_variable` 
```

> and some that does not contain a `dependent_variable`, then training is 

Reviewer 2: Think this repetition can be omitted:
```suggestion
and some that do not, the model is trained
```

> performed on the `training_percent` of the documents that contain ground truth 

Reviewer 2: ```suggestion
on the `training_percent` of the documents that contain ground truth. However,
```

> and predictions will be made against all of the data. The current implementation 

Reviewer 2: Changing to present tense:
```suggestion
predictions are made against all of the data. The current implementation 
```

of {reganalysis} supports a single batch analysis for both training and 

> predictions. Future versions will allow a learned model to be incorporated at 

Reviewer 2: > Future versions will ...

I generally try to avoid putting promises like this in the docs in case they fall through.
